### PR TITLE
Add type hints

### PR DESCRIPTION
**Pull request by thtrummer**
This is an attempt to add type hints to the library. Inferring types from an existing code base is somewhat tricky so some thorough review is required. This is compatible with Python 3.5 to not cut of too many people (that would also include myself).

### CODE REVIEW

File: ifaddr/_shared.py [hunk 1]
```diff
@@ -69,7 +75,13 @@ class IP(object):
     Represents an IP address of an adapter.
     """
 
-    def __init__(self, ip, network_prefix, nice_name):
+    # Type of an IPv4 address (a string in "xxx.xxx.xxx.xxx" format)
+    IPv4Address = Union[str]
```
Comment (jstasiak) [inline ifaddr/_shared.py:79]: Let's put those at module level and prefix with `_` to not suggest those are to be used by the client code.

File: ifaddr/_posix.py [hunk 2]
```diff
@@ -62,18 +66,20 @@ def add_ip(adapter_name, ip):
...
+        ip_addr = shared.sockaddr_to_ip(addr[0].ifa_addr)
+        if ip_addr:
             if addr[0].ifa_netmask and not addr[0].ifa_netmask[0].sa_familiy:
                 addr[0].ifa_netmask[0].sa_familiy = addr[0].ifa_addr[0].sa_familiy
             netmask = shared.sockaddr_to_ip(addr[0].ifa_netmask)
             if isinstance(netmask, tuple):
                 netmaskStr = str(netmask[0])
                 prefixlen = shared.ipv6_prefixlength(ipaddress.IPv6Address(netmaskStr))
-            else:
+            elif isinstance(netmask, str):
```
Comment (jstasiak) [inline ifaddr/_posix.py:77]: Please don't change runtime behavior (even theoretically). If you have some mypy-related issues in this block use casts (if in doubt let me know).

File: ifaddr/_posix.py [hunk 3]
```diff
@@ -82,4 +88,4 @@ def add_ip(adapter_name, ip):
 
     libc.freeifaddrs(addr0)
 
-    return ips.values()
+    return list(ips.values())
```
Comment (jstasiak) [inline ifaddr/_posix.py:91]: No changing runtime behavior please, please don't change the return type in a PR that's supposed to document the *existing* behavior.

File: ifaddr/_shared.py [hunk 4]
```diff
@@ -35,15 +41,15 @@ class Adapter(object):
...
+    def __init__(self, name: str, nice_name: str, ips: List['IP'], index: Optional[int] = None) -> None:
 
         #: Unique name that identifies the adapter in the system.
         #: On Linux this is of the form of `eth0` or `eth0:1`, on
         #: Windows it is a UUID in string representation, such as
         #: `{846EE342-7039-11DE-9D20-806E6F6E6963}`.
         self.name = name
 
-        #: Human readable name of the adpater. On Linux this
+        #: Human readable name of the adapter. On Linux this
```
Comment (jstasiak) [inline ifaddr/_shared.py:52]: No unrelated changes please.

File: ifaddr/_shared.py [hunk 5]
```diff
@@ -143,25 +155,25 @@ class sockaddr_in6(ctypes.Structure):
...
                    ('sin_addr', ctypes.c_uint8 * 4),
                    ('sin_zero', ctypes.c_uint8 * 8)]
 
-    class sockaddr_in6(ctypes.Structure):
+    class sockaddr_in6(ctypes.Structure):  # type: ignore
         _fields_= [('sin6_familiy', ctypes.c_uint16),
                    ('sin6_port', ctypes.c_uint16),
                    ('sin6_flowinfo', ctypes.c_uint32),
                    ('sin6_addr', ctypes.c_uint8 * 16),
                    ('sin6_scope_id', ctypes.c_uint32)]
```
Comment (jstasiak) [inline ifaddr/_shared.py:174]: Why re those '#type: ignore' annotations above needed? (curious)

File: ifaddr/_shared.py [hunk 6]
```diff
@@ -69,7 +75,13 @@ class IP(object):
     Represents an IP address of an adapter.
     """
 
-    def __init__(self, ip, network_prefix, nice_name):
+    # Type of an IPv4 address (a string in "xxx.xxx.xxx.xxx" format)
+    IPv4Address = Union[str]
+
+    # Type of an IPv6 address (a three-tuple `(ip, flowinfo, scope_id)`)
+    IPv6Address = Union[Tuple[str, int, int]]
+
```
Comment (jstasiak) [inline ifaddr/_shared.py:83]: `Unions` with single items aren't making much sense, just use `str` and `Tuple[str, int, int]` above.

File: ifaddr/_win32.py [hunk 7]
```diff
@@ -102,10 +106,10 @@ def get_adapters(include_unconfigured=False):
                                       ctypes.byref(addressbuffer),
                                       ctypes.byref(addressbuffersize))
     if retval != NO_ERROR:
-        raise ctypes.WinError()
+        raise ctypes.WinError()  # type: ignore
```
Comment (jstasiak) [inline ifaddr/_win32.py:109]: What's the error being silenced here?

File: ifaddr/_win32.py [hunk 8]
```diff
@@ -84,11 +87,12 @@ def enumerate_interfaces_of_adapter(nice_name, address):
 
     for address in addresses:
         ip = shared.sockaddr_to_ip(address.Address.lpSockaddr)
-        network_prefix = address.OnLinkPrefixLength
-        yield shared.IP(ip, network_prefix, nice_name)
+        if ip:
+            network_prefix = address.OnLinkPrefixLength
+            yield shared.IP(ip, network_prefix, nice_name)
```
Comment (jstasiak) [inline ifaddr/_win32.py:92]: Please don't change the code behavior.

**Review by jstasiak — CHANGES_REQUESTED** [review]
Looks good in general, thanks for the contribution. Few things need to be changed.

File: ifaddr/_shared.py (hunk 6 above)
Comment (thtrummer) [reply ifaddr/_shared.py:83]: This is actually correct however it will be unnecessary when moved outside the class.

File: ifaddr/_win32.py (hunk 7 above)
Comment (thtrummer) [reply ifaddr/_win32.py:109]: The Linux version of `ctypes` doesn't contain `WinError`.

File: ifaddr/_shared.py (hunk 5 above)
Comment (thtrummer) [reply ifaddr/_shared.py:174]: `Mypy` expects types to have a single definition (and ignores platform-dependent `if` statements).

File: ifaddr/_posix.py (hunk 3 above)
Comment (thtrummer) [reply ifaddr/_posix.py:91]: I will change this back, however, leaking implementation details needs to be addressed at some point.

File: ifaddr/_posix.py (hunk 2 above)
Comment (thtrummer) [reply ifaddr/_posix.py:77]: I've changed this back, however, the non-existing error handling is an issue (which can be fixed later I guess).

Comment (jstasiak) [reply ifaddr/_posix.py]: Sure, if there's anything missing here I'm happy to accept a patch improving things. I don't think we can get `None` here but maybe I'm wrong.

File: ifaddr/_posix.py (hunk 3 above)
Comment (jstasiak) [reply ifaddr/_posix.py]: It may or may not need to be addressed – it we document this return types as `Iterable` and explicitly mention in the documentation that static type hints can be depended on while the actual types returned (which can be narrower) shouldn't, I think we're fine here.

File: ifaddr/_posix.py [hunk 9]
```diff
@@ -36,9 +38,9 @@ class ifaddrs(ctypes.Structure):
                     ('ifa_addr', ctypes.POINTER(shared.sockaddr)),
                     ('ifa_netmask', ctypes.POINTER(shared.sockaddr))]
 
-libc = ctypes.CDLL(ctypes.util.find_library("socket" if os.uname()[0] == "SunOS" else "c"), use_errno=True)
+libc = ctypes.CDLL(ctypes.util.find_library("socket" if os.uname()[0] == "SunOS" else "c"), use_errno=True)  # type: ignore
 
-def get_adapters(include_unconfigured=False):
+def get_adapters(include_unconfigured: bool = False) -> 'collections._OrderedDictValuesView':
```
Comment (jstasiak) [inline ifaddr/_posix.py:43]: I think I'll change this to just `Iterable[...]` but I'll merge as is for now.

**Comment by jstasiak** [comment]
Thank you once again, this is very useful! I recently dropped Python 3.5 support (it's not released yet) so keep in mind that there's no guarantee that the type hints will remain 3.5-compatible.

**Review by jstasiak — APPROVED** [review]

File: ifaddr/_posix.py (hunk 9 above)
Comment (thtrummer) [reply ifaddr/_posix.py:43]: That's actually not a bad idea. In that case it makes probably sense to change it for the Windows version as well, so both have the same type signature (since `Mypy` is currently unhappy about that).
